module/zfs/dmu_recv.c

```diff
@@ -1333,16 +1346,34 @@ receive_object(struct receive_writer_arg *rwa, struct drr_object *drro,
 	/* handle more restrictive dnode structuring for raw recvs */
 	if (rwa->raw) {
 		/*
-		 * Set the indirect block shift and nlevels. This will not fail
-		 * because we ensured all of the blocks were free earlier if
-		 * this is a new object.
+		 * Set the indirect block size, block shift, nlevels.
+		 * This will not fail because we ensured all of the
+		 * blocks were freed earlier if this is a new object.
+		 * For non-new objects block size and indirect block
+		 * shift cannot change and nlevels can only increase.
 		 */
 		VERIFY0(dmu_object_set_blocksize(rwa->os, drro->drr_object,
 		    drro->drr_blksz, drro->drr_indblkshift, tx));
 		VERIFY0(dmu_object_set_nlevels(rwa->os, drro->drr_object,
 		    drro->drr_nlevels, tx));
+
+		/*
+		 * Set the maxblkid. We will never free the first block of
+		 * an object here because a maxblkid of 0 could indicate
+		 * an object with a single block or one with no blocks.
+		 * If we are reducing the maxblkid we must first free
+		 * everything above that blkid to ensure the DMU is
+		 * consistent with itself.
+		 */
+		err = dmu_free_long_range(rwa->os, drro->drr_object,
+		    (drro->drr_maxblkid + 1) * drro->drr_blksz,
+		    DMU_OBJECT_END);
```

Comment:
I'm pretty sure you can not call dmu_free_long_range when the thread has already assigned a different tx.  It may deadlock because dmu_free_long_range() will assign a new tx, which may need to wait for a txg to sync, which needs to wait for the current tx to be committed.

Maybe we can do this before assigning the tx?  You might also mention in the comment that in this case there will be a FREE record telling us to free everything after the maxblkid, but we need to do it earlier so that the range is marked as to-be-freed before dmu_object_set_maxblkid() decreases the maxblkid.